radix-engine/src/blueprints/access_controller/package.rs

```diff
@@ -565,9 +705,134 @@ impl AccessControllerNativePackage {
             address,
         )?;
 
+        // Invoking the post-initialization method on the component
+        api.call_method(
+            &node_id,
+            ACCESS_CONTROLLER_POST_INSTANTIATION_IDENT,
+            scrypto_encode(&AccessControllerPostInstantiationInput).unwrap(),
+        )?;
+
         Ok(IndexedScryptoValue::from_typed(&address))
     }
 
+    /// This method is only callable when the access controller virtual direct package caller
+    /// badge is present.
+    ///
+    /// This method has been added due to an issue with setting the metadata of the recovery badge
+    /// resource to include the address of the access controller before the access controller was
+    /// globalized. Doing that lead to an error along the lines of "callframe has no reference to
+    /// the (access controller) node id" because the access controller's node id has been allocated
+    /// but nothing has been globalized to it. Thus, the call frame did not have a reference to the
+    /// global address to pass when calling the metadata set method.
+    ///
+    /// A minimal example that reproduces this issue is only possible after the metadata interfaces
+    /// are updated to support any MetadataEntry.
+    ///
+    /// The method below can be removed and the logic can be moved to the instantiation function
+    /// once this bug is fixed.
+    fn post_instantiation<Y>(
+        input: &IndexedScryptoValue,
+        api: &mut Y,
+    ) -> Result<IndexedScryptoValue, RuntimeError>
+    where
+        Y: ClientApi<RuntimeError>,
+    {
+        input
+            .as_typed::<AccessControllerPostInstantiationInput>()
+            .map_err(|e| {
+                RuntimeError::SystemUpstreamError(SystemUpstreamError::InputDecodeError(e))
+            })?;
+
+        let global_address = api.actor_get_global_address()?;
+        let resource_address = {
+            let substate_key = AccessControllerField::AccessController.into();
+            let handle =
+                api.actor_lock_field(OBJECT_HANDLE_SELF, substate_key, LockFlags::read_only())?;
+
+            let access_controller = {
+                let access_controller: AccessControllerSubstate =
+                    api.field_lock_read_typed(handle)?;
+                access_controller
+            };
+            access_controller.recovery_badge
+        };
+
+        api.call_module_method(
+            resource_address.as_node_id(),
+            ObjectModuleId::Metadata,
+            METADATA_SET_IDENT,
+            scrypto_encode(&MetadataSetInput {
+                key: "name".to_owned(),
+                value: scrypto_decode(
+                    &scrypto_encode(&MetadataEntry::Value(MetadataValue::String(
+                        "Recovery Badge".into(),
+                    )))
+                    .unwrap(),
+                )
+                .unwrap(),
+            })
+            .unwrap(),
+        )?;
```

Comment:
Can leverage the Native SDK to simplify the module method call?

Reply:
Sure can do! I initially didn't use the native sdk as I thought we'd soon fix the issue anyway. I've updated the PR to use the native SDK here :)